cliapi/cliapi.py

```diff
@@ -0,0 +1,257 @@
+        'query=': 'specify a python dictionary style query command',
+        'all': 'output all API results for specified API options or defaults',
+    }
+
+    if provider == None:
+        vpp = Provider()
+        # no provider supplied so at least print out the "common" help...
+        vpp.help = help
+    else:
+        vpp = valid_providers[provider].provider
+        # update help with options from this provider...
+        vpp.help.update(help)
+    required = list()
+    for fetcher in vpp.fetchers.items():
+        # create a list of required options for fast lookup...
+        required += fetcher[1][1][0]
+    indent2_format = '  --{:<15}  {:<15}'
+    print("usage: {} [display option#1]... [API option#1]... [CLI option]".format(sys.argv[0]))
+    print("\n***[ {} ]*** provider Display options:".format(provider))
+    for key in vpp.scoops.keys():
+        # list display options for this provider...
+        print (indent2_format.format(key, vpp.help.get(key, '')))
+    print("\n***[ {} ]*** provider API config options:".format(provider))
+    for k, v in vpp.options.items():
+        # list API configuration options for this provider...
+        if v.startswith('$'):
+            opt = v[1:]
+            help = vpp.help.get(opt,'')
+            if help != '':
+                # help message is provided by plugin provider...
+                help = ''.join((', ', help) )
+            if v in required:
+                # if this option is required, then say so...
+                help = ''.join(('-REQUIRED-', help))
+            elif opt in vpp.template:
+                # otherwise display the default...
+                help = ''.join(('default=\'', vpp.template[opt], '\'', help))
+
+            print (indent2_format.format(opt + '=', help))
+    print("\nCommon CLI options:")
+    for key in cli_options:
+        # list of common CLI options supported by this module, across all providers
+        print (indent2_format.format(key, vpp.help.get(key, '')))
+
+    exit(-1)
+
+
+def _sandbox_eval(vpp, lookup):
+    return eval('vpp' + lookup,
+                {'__builtins__': None},
+                {'vpp': vpp})
+
+
+def main():
+    # must do our own parsing here since we don't know the
+    # valid options until we establish the plugin provider
+    # -- getopt() does not allow this usage...
+    ct = _cli_parse(sys.argv[1:])
+
+    if '--list-providers' in ct:
+        print(json.dumps(list(valid_providers.keys()), indent=2))
+        exit(0)
+
+    cmd_dict = {}
+    if '--provider' in ct:
+        # using supplied provider...
+        provider = ct['--provider']
+    else:
+        try:
+            # use the first valid provider in list...
+            provider = list(valid_providers.keys())[0]
+        except Exception as e:
+            # error -- print help and exit
+            print('No plugin providers found')
+            _print_help(None)
+            exit(-1)
+
```

Comment:
For the general hard coded options we should probably take advantage of argparse or some other existing CLI parsing framework that should reduce some of the work that needs to be done in _cli_parse() and will also allow us to handle short and long options in a "standard" way.